### PR TITLE
[#132] Deploy Mode

### PR DESCRIPTION
**Pull request by MichaelBrew**
Closes #606 
Feature request: https://github.com/helium/hotspot-app/issues/132
Designs: https://www.figma.com/file/v6nnssiwAaVWEvkkSmK7Uq/Deploy-Mode-Designs?node-id=0%3A110

## Summary
Add support for "Deploy Mode" in Settings. Enabling "Deploy Mode" will permanently disable the ability to view secret words, transfer hotspots, and send any payments unless the user signs out and imports their account again. However, the user can specify a whitelisted, permanent payment address when enabling "Deploy Mode" to allow payments to that address only (if no address was entered, then no payments are allowed).

When opening the "Send" screen now, there are three primary scenarios:
- **Deploy Mode disabled (default)**: No change, Send screen should work as it always has
- **Deploy Mode enabled with permanent address**: Address input is prefilled and locked, QR scan option is removed
- **Deploy Mode enabled without permanent address**: Amount and address inputs are empty and locked, grey "Disabled" banner displayed above inputs, QR scan option is removed, and red "Payments are disabled in Deploy Mode" text is shown below submit button

## Screenshots
Demo: https://www.dropbox.com/s/d6xhukysrhiof2t/deploy_mode_demo.mov?dl=0
| Settings (disabled) | Confirmation | Settings (enabled) | Send View (w/ address) | Send View (no address) | Transfer Hotspot |
| ---- | ---- | ---- | ---- | ---- | ---- |
| Snip20210801_41 | image | Snip20210801_44 | Snip20210801_45 | Snip20210801_46 | Snip20210814_24 |

## Implementation
"Deploy Mode" relies on the `secureAccount` util to enable, disable, and restore state from the local device. The secure item is called `deployModeEnabled`, and is imported and subsequently managed by `appSlice`.

Regarding the permanent payment address, the "Submit" button on `DeployModeModal` is enabled by default, but will become disabled if any input exists in "Address" field that is not a valid wallet address. Upon submission, the address is stored to secure items as `permanentPaymentAddress`. That is then fetched when rendering `SendScreen` and passed down to `SendView` (and eventually to `SendForm` and `SendDetailsForm`).

The `SendView` component (and its relevant children) now accepts three new props: `isDisabled` (completely locks the form), `lockedPaymentAddress` (only locks address field with prefilled input), and `warning` (text disabled underneath submission button in red). The `SendLockedHeader` component can now accept dynamic text/colors which allows us to show a blue "QR Info" banner after scanning a QR code, or a grey "Disabled" banner if Deploy Mode is enabled without a payment address.

This PR also introduces a `HeliumBottomSheet` component that is used by both `HeliumActionSheet` and the new `DeployModeModal`. It manages the animation/blurring/container styles, and renders the components children as its 
 content.

## i18n
The following strings have been added and translated to Japanese, Korean, and simplified Chinese (with the help of Google Translate):
- `generic.disabled`: 'Disabled'
- `generic.one`: 'one'
- `generic.swipe_to_confirm`: 'Swipe to Confirm'
- `send.deployModePaymentsDisabled`: 'Payments are disabled in Deploy Mode'
- `deployMode.title`: 'Deploy Mode',
- `deployMode.subtitle`: 'This mode adds extra protection to your wallet, restricting some app features.',
- `deployMode.inDeployMode`: 'In Deploy Mode:'
- `deployMode.cantViewWords`: "Can't view your 12 secure words"
- `deployMode.cantTransferHotspots`: "Can't transfer Hotspots from this account"
- `deployMode.canOnlySendFunds`: 'Can only send funds to'
- `deployMode.otherAccount`: 'other specified account'
- `deployMode.enableButton`: 'Enable Deploy Mode'
- `deployMode.disableInstructions`: 'In order to disable this feature, you will have to log out. Remember to write down all 12 words now.'
- `deployMode.addressLabel`: 'Allowed Account Address...'
- `transfer.deployModeTransferDisableTitle`: 'Transfer Hotspot Disabled'
- `transfer.deployModeTransferDisabled`: 'Transfer Hotspot is disabled while in Deploy Mode.'

## Testing
So far these changes have been manually tested by enabling "Deploy Mode", asserting the expected UI and inability to reveal secret words. Killing the app in simulator and re-opening also shows that "Deploy Mode" is still enabled across sessions (as expected). Also tested the language option to ensure the HeliumActionSheet still behaves as expected (with the new HeliumBottomSheet). Below are some test cases for the "Send" view:
- ✅ Entering some input into Deploy Mode address field, then closing modal without submission, then re-opening modal yields an empty address field
- ✅ Deploy Mode disabled, send payment to another wallet address
- ✅ Deploy Mode enabled with address, send payment to whitelisted address
- ✅  Deploy Mode enabled with address, whitelisted address persists across sessions (kill app, re-open, view pre-filled address in "Send" form)
- ✅ Deploy Mode enabled without address, unable to send any payments
- ✅ Deploy Mode enabled without address, state persists across sessions (kill app, re-open, view disabled "Send" form)
- ✅ Logging out and re-importing account resets all Deploy Mode state

## TODO
- [x] Update translations
- [x] Fix Deploy Mode screen going out of modal when opening keyboard
- [x] Decide on "Deploy Mode Enabled" copy on splash screen
  - Decision is to omit splash screen changes due to inherent complexity: https://discord.com/channels/404106811252408320/869618232632410172/883785391222693899
- ~Add arrow asset to "Swipe to Confirm" button~
  - Not implemented due to ambiguity with arrow behavior as user swipes thumb

### CODE REVIEW

**Comment by MichaelBrew** [comment]
Apologies for all the old commits from a previous PR - I branched off that before it was merged to main, but the file diff here looks accurate

**Comment by cokes518** [comment]
Hey @MichaelBrew nice PR! Looking at the rationale behind the feature request, since it was written in February and seems like a lifetime ago, we've actually added a lot of host/owner features that seems to negate the need for a deployment mode (as spec'd out in this PR, with the intention to just hide 12 words). 

As an example, there's no need to pass on the app to a host to assert (and thus hide the words) since there's now remote assert support.

There's also no need to pass the phone to a host to run diagnostics (and thus to hide the words) since hosts can follow a hotspot and run diagnostics on their own phone.

I think the key part of deployment mode then, is restricting damaging and irreversible actions *in addition to revealing the 12 words*. Enabled deployment mode doesn't matter much if you can still empty the wallet and transfer all the hotspots out of the account.

We can scope out this PR to include restricting payment transactions (any payment transaction that debits the account) and hotspot transfers, and renaming it to "Secure Mode", it can capture the spirit of deployment mode a bit more, with additional security features in mind. 

What do you think?

**Comment by tyler-whitman** [comment]
Thanks for submitting another great change! We will get to this review soon, just finishing up some changes for our next release and then we can get this in for the following one I think.

**Comment by MichaelBrew** [comment]
Hey @cokes518 just want to check back in and see if you have any more thoughts about this PR and https://github.com/MichaelBrew/hotspot-app/pull/2? Open to feedback if we want to tweak any of presentation / feature language / functionality.

**Comment by cokes518** [comment]
> Hey @cokes518 just want to check back in and see if you have any more thoughts about this PR and [MichaelBrew#2](https://github.com/MichaelBrew/hotspot-app/pull/2)? Open to feedback if we want to tweak any of presentation / feature language / functionality.

hey @MichaelBrew! I'd love to see #2 merged with this one and shipped as a single feature.

**Comment by MichaelBrew** [comment]
@cokes518 ok sure thing, I'll go ahead and merge those into a single PR here 👍

**Comment by cokes518** [comment]
@MichaelBrew We're doing the final reviews and it seems there is a feature-gap where a user in deployment mode can transfer Hotspots out of the account. 

Ideally if the app is in this mode, we also restrict their ability to transfer Hotspot (either all together, or to the allowed account address for payments).

Otherwise, this feature seems easy to work around and is not really secure in its intended implementation. Would like your thoughts on this.

**Review by tyler-whitman — APPROVED** [review]
Code looks good and it works as expected. I didn't get to test everything, but it looks like payment / QR is working as expected still and disabled properly when in deploy mode. I did find one UI glitch, when you enter allowed account addresses when enabling deploy mode on iOS the text shifts out of the modal container.
@cokes518 if you are ok with it code looks good. Maybe @matthewcarlreetz wants to take another look at the code before we merge it

File: src/features/moreTab/more/MoreListItem.tsx [hunk 1]
```diff
@@ -60,6 +72,11 @@ const MoreListItem = ({
     [],
   )
 
+  let textColor = 'primaryText'
```
Comment (tyler-whitman) [inline src/features/moreTab/more/MoreListItem.tsx:75]: This needs a type to fix lint errors:

` let textColor: Colors = 'primaryText'`

**Review by tyler-whitman — CHANGES_REQUESTED** [review]

**Comment by tyler-whitman** [comment]
Closed with https://github.com/helium/hotspot-app/pull/948

**Pull request closed without merging**